### PR TITLE
Improvements in decompiler output

### PR DESCRIPTION
**Pull request by siegfriedpammer**
* Add header region to decompiled code (useful for debugging).
* Add EscapeInvalidIdentifiers transform to make sure generated code is always parsable.

Note that this would make diagnosing things like #25302 a lot easier.

<details><summary>Ask Mode template not completed</summary>

### Customer scenario

* It is useful to see where the decompiled code comes from.
* Also adding the EscapeInvalidIdentifiers transform allows the parser to work with code that contains compiler-generated or invalid identifiers.

### Bugs this fixes

* change mentioned in https://github.com/dotnet/roslyn/issues/25246#issuecomment-372064351

### Workarounds, if any

Unknown.

### Risk

This is low risk, because the only change that is complex is the `EscapeInvalidIdentifiers` transform, which has already been tested well during the development and testing of ILSpy.

### Performance impact

Low performance impact, because the EscapeInvalidIdentifiers transform is very simple.

### Root cause analysis

How did we miss it?  What tests are we adding to guard against it in the future?

### How was the bug found?

(E.g. customer reported it vs. ad hoc testing)

### Test documentation updated?

If this is a new non-compiler feature or a significant improvement to an existing feature, update https://github.com/dotnet/roslyn/wiki/Manual-Testing once you know which release it is targeting.

</details>

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs [hunk 1]
```diff
@@ -224,15 +224,28 @@ private string GetRootPathWithGuid_NoLock()
...
+            // (This happens for example, when there is compiler-generated code that is not yet recognized/transformed by the decompiler.)
+            decompiler.AstTransforms.Add(new ICSharpCode.Decompiler.CSharp.Transforms.EscapeInvalidIdentifiers());
+
             var fullTypeName = new FullTypeName(fullName);
 
+            var decompilerVersion = System.Diagnostics.FileVersionInfo.GetVersionInfo(typeof(CSharpDecompiler).Assembly.Location);
+
+            // Add header to match output of metadata-only view.
+            // (This also makes debugging easier, because you can see which assembly was decompiled inside VS.)
+            var header = $"#region Assembly {assemblyDefinition.FullName}" + Environment.NewLine
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs]: Per here:

https://github.com/dotnet/roslyn/blob/e123a4207123d5df8c45fb1caba0fb3acbae0e00/src/Features/Core/Portable/MetadataAsSource/MetadataAsSourceHelpers.cs#L45

it looks like we localize "Assembly". You should be able to access FeaturesResources from here so just a quick change.

Comment (siegfriedpammer) [reply src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs]: Thank you for the hint! It's fixed now.

File: src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs [hunk 2]
```diff
@@ -224,15 +224,28 @@ private string GetRootPathWithGuid_NoLock()
...
             // Load the assembly.
-            var ad = AssemblyDefinition.ReadAssembly(assemblyLocation, new ReaderParameters() { AssemblyResolver = new RoslynAssemblyResolver(compilation) });
+            var assemblyDefinition = AssemblyDefinition.ReadAssembly(assemblyLocation, new ReaderParameters() { AssemblyResolver = new RoslynAssemblyResolver(compilation) });
 
             // Initialize a decompiler with default settings.
-            var decompiler = new CSharpDecompiler(ad.MainModule, new DecompilerSettings());
+            var decompiler = new CSharpDecompiler(assemblyDefinition.MainModule, new DecompilerSettings());
+            // Escape invalid identifiers to prevent Roslyn from failing to parse the generated code.
+            // (This happens for example, when there is compiler-generated code that is not yet recognized/transformed by the decompiler.)
+            decompiler.AstTransforms.Add(new ICSharpCode.Decompiler.CSharp.Transforms.EscapeInvalidIdentifiers());
```
Comment (sharwell) [inline src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs]: ❓ Can we add an import and simplify this construction?

File: src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs [hunk 3]
```diff
@@ -224,15 +224,28 @@ private string GetRootPathWithGuid_NoLock()
...
+
             var fullTypeName = new FullTypeName(fullName);
 
+            var decompilerVersion = System.Diagnostics.FileVersionInfo.GetVersionInfo(typeof(CSharpDecompiler).Assembly.Location);
```
Comment (sharwell) [inline src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs]: ❓ Can we add an import and simplify this construction?

**Comment by sharwell** [comment]
@Pilchie Bar check for 15.7

**Comment by Pilchie** [comment]
Yes, I would take this for 15.7.

**Comment by sharwell** [comment]
@siegfriedpammer This needs to be rebased onto the **dev15.7.x** branch so we can take it for an earlier release than the **master** branch. I'm happy to make the change for you if you don't want to do it yourself. Let me know! 😄

**Comment by siegfriedpammer** [comment]
@sharwell Yes, it would be nice, if you could do that for me. Thank you very much! 😄

**Comment by sharwell** [comment]
@siegfriedpammer Done, thanks!

**Comment by Pilchie** [comment]
Approved for 15.7 Preview 3.

**Comment by sharwell** [comment]
@dotnet-bot retest windows_coreclr_debug_prtest please

**Pull request merged**